Hardware credential authentication

chaotixkilla opened the pull request.

chaotixkilla committed:
feat: implement hardware key auth

chaotixkilla commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add fido2 auth protocol support](https://trello.com/c/5sqivfNl/491-add-fido2-auth-protocol-support)

hellypi commented on lib/hardware-credentials.js:
```diff
@@ -0,0 +1,36 @@
+}
+
+function getHardwareCredentials () {
+  const sql = `SELECT * from hardware_credentials`
```
```suggestion
  const sql = `SELECT * FROM hardware_credentials`
```

hellypi commented on lib/hardware-credentials.js:
```diff
@@ -0,0 +1,36 @@
+}
+
+function getHardwareCredentialsOfUser (userID) {
+  const sql = `SELECT * from hardware_credentials where user_id=$1`
```
```suggestion
  const sql = `SELECT * FROM hardware_credentials WHERE user_id=$1`
```

hellypi commented on lib/new-admin/graphql/modules/authentication/FIDO2FAStrategy.js:
```diff
@@ -0,0 +1,165 @@
+const REMEMBER_ME_AGE = 90 * T.day
+
+const rpID = `localhost`
+const expectedOrigin = `https://${rpID}:3001`
```
shouldn't this be an environment variable? Seems like it's gonna break on something like digitalocean

hellypi commented on lib/new-admin/graphql/modules/authentication/FIDO2FAStrategy.js:
```diff
@@ -0,0 +1,165 @@
+const generateAttestationOptions = (userID, session) => {
+  return users.getUserById(userID).then(user => {
+    return Promise.all([credentials.getHardwareCredentialsOfUser(user.id), user])
+  }).then(([userDevices, user]) => {
```
according to pg-promise documentation, chained queries should be done in the same task (no real need for a full on transaction here).
"Task" allows us to reuse the same db connection for chained queries instead of potentially doing something like
reserve connection --> query --> drop connection back to the pool --> reserve connection --> query -->...

https://vitaly-t.github.io/pg-promise/Task.html

hellypi commented on lib/new-admin/graphql/modules/authentication/FIDO2FAStrategy.js:
```diff
@@ -0,0 +1,165 @@
+const generateAssertionOptions = (username, password, context) => {
+  return userManagement.authenticateUser(username, password).then(user => {
+    return credentials.getHardwareCredentialsOfUser(user.id).then(devices => {
+      const options = simpleWebauthn.generateAssertionOptions({
```
these are also chained queries, right? If so, try to see if you can also use a task here.

hellypi commented on lib/new-admin/graphql/modules/authentication/FIDO2FAStrategy.js:
```diff
@@ -0,0 +1,165 @@
+      expectedChallenge: `${expectedChallenge}`,
+      expectedOrigin,
+      expectedRPID: rpID
+    }).then(async verification => {
```
I personally also prefer async await where possible, but check with Rafael if using async here won't break the thing we have about not using async in lamassu-server

hellypi replied:
you don't seem to need the response of users.getUserById to do simpleWebauthn.verifyAttestationResponse so we can create a promise array and do Promise.all([promise1, promise2])

hellypi commented on lib/new-admin/graphql/modules/authentication/FIDO2FAStrategy.js:
```diff
@@ -0,0 +1,165 @@
+  return userManagement.authenticateUser(username, password).then(user => {
+    const expectedChallenge = context.req.session.webauthn.assertion.challenge
+
+    return credentials.getHardwareCredentialsOfUser(user.id).then(async devices => {
```
same here about async

hellypi commented on lib/new-admin/graphql/modules/authentication/FIDO2FAStrategy.js:
```diff
@@ -0,0 +1,165 @@
+    }).then(async verification => {
+      const { verified, attestationInfo } = verification
+
+      if (verified && attestationInfo) {
```
can't we invert this? if (!(verified || attestationInfo)) return // response for when not verified or atested
// rest of the code

hellypi commented on lib/new-admin/graphql/modules/authentication/FIDOPasswordlessStrategy.js:
```diff
@@ -0,0 +1,164 @@
+const REMEMBER_ME_AGE = 90 * T.day
+
+const rpID = `localhost`
+const expectedOrigin = `https://${rpID}:3001`
```
same thing about env vars

hellypi commented on lib/new-admin/graphql/modules/authentication/FIDOPasswordlessStrategy.js:
```diff
@@ -0,0 +1,164 @@
+  return users.getUserById(userID).then(user => {
+    return Promise.all([credentials.getHardwareCredentialsOfUser(user.id), user])
+  }).then(([userDevices, user]) => {
+    const options = simpleWebauthn.generateAttestationOptions({
```
try using db.task for chained queries

hellypi commented on lib/new-admin/graphql/modules/authentication/FIDOPasswordlessStrategy.js:
```diff
@@ -0,0 +1,164 @@
+
+const generateAssertionOptions = (username, context) => {
+  return users.getUserByUsername(username).then(user => {
+    return credentials.getHardwareCredentialsOfUser(user.id).then(devices => {
```
db.task

hellypi commented on lib/new-admin/graphql/modules/authentication/FIDOPasswordlessStrategy.js:
```diff
@@ -0,0 +1,164 @@
+  return users.getUserByUsername(username).then(user => {
+    const expectedChallenge = context.req.session.webauthn.assertion.challenge
+
+    return credentials.getHardwareCredentialsOfUser(user.id).then(async devices => {
```
db task

hellypi commented on lib/new-admin/graphql/modules/userManagement.js:
```diff
@@ -141,10 +143,14 @@ const deleteSession = (sessionID, context) => {
-  })
+  return authenticateUser(username, password)
+    .then(user => {
+      return Promise.all([credentials.getHardwareCredentialsOfUser(user.id), user.twofa_code])
```
you get the idea, for the rest of this review whenever you find chained queries, try to see if you can do it using db.task

hellypi commented on migrations/1620335170327-hardware-credentials.js:
```diff
@@ -0,0 +1,20 @@
+
+exports.up = function (next) {
+  var sql = [
+    `ALTER TABLE user_register_tokens ADD COLUMN use_fido BOOLEAN DEFAULT false`,
+    `CREATE TABLE hardware_credentials (
+      id UUID PRIMARY KEY,
+      user_id UUID REFERENCES users(id),
+      created TIMESTAMPTZ DEFAULT now(),
+      last_used TIMESTAMPTZ DEFAULT now(),
+      data JSONB
```
maybe make these fields not nullable

hellypi commented on new-lamassu-admin/src/pages/Authentication/LoginState.js:
```diff
@@ -71,6 +97,43 @@ const LoginState = ({ state, dispatch }) => {
+    GENERATE_ASSERTION,
+    {
+      onCompleted: ({ generateAssertionOptions: options }) => {
+        console.log(options)
```
stray console.log

hellypi commented on new-lamassu-admin/src/pages/UserManagement/UserManagement.js:
```diff
@@ -62,6 +81,26 @@ const Users = () => {
 
+  const [validateAttestation] = useMutation(VALIDATE_ATTESTATION, {
+    onCompleted: res => {
+      console.log(res)
+      // success ? console.log('success') : console.log('failure')
```
stray console.logs

chaotixkilla committed:
fix: sql uppercasing
fix: structural changes to remove async/await flows
fix: invert boolean flow
fix: minor fixes

josepfo commented on new-lamassu-admin/src/pages/Authentication/InputFIDOState.js:
```diff
@@ -0,0 +1,214 @@
+              username: localClientField
+            },
+      onCompleted: ({ generateAssertionOptions: options }) => {
+        console.log(options)
```
remove console.log

chaotixkilla committed:
fix: remove log

RafaelTaranto commented on new-lamassu-admin/src/pages/UserManagement/UserManagement.js:
```diff
@@ -62,6 +81,25 @@ const Users = () => {
   const [userInfo, setUserInfo] = useState(null)
 
+  const [validateAttestation] = useMutation(VALIDATE_ATTESTATION, {
+    onCompleted: res => {
+      // TODO: show a brief popup to have UX feedback?
+    }
```
What does actually happen here? Does the user get logged in?

RafaelTaranto commented on lib/hardware-credentials.js:
```diff
@@ -0,0 +1,36 @@
+  return db.any(sql)
+}
+
+function getHardwareCredentialsOfUser (userID) {
```
`getHardwareCredentialsByUserId`

RafaelTaranto commented on lib/new-admin/graphql/resolvers/users.resolver.js:
```diff
@@ -1,34 +1,81 @@
+      }
+    default:
+      return {}
+  }
```
Can't we use a switch to setup the options instead of everything?

Something like:
```js
     case 'FIDO2FA': {
        return { username, password }
     }
```

Also if you change the generateAssertionOptions you can always pass down all params and have the same code for everything. Actually since the options are the only thing that chance maybe generateAssertion(context, options) would be a better interface for the functions.

RafaelTaranto commented on lib/new-admin/graphql/resolvers/users.resolver.js:
```diff
@@ -1,34 +1,81 @@
+        validateAssertion: (...[, { assertionResponse }, context]) => authentication.strategy.validateAssertion(assertionResponse, context)
+      }
+    default:
+      return {}
```
same as above

RafaelTaranto commented on lib/new-admin/graphql/modules/authentication/FIDOUsernamelessStrategy.js:
```diff
@@ -0,0 +1,186 @@
+const REMEMBER_ME_AGE = 90 * T.day
+
+const rpID = devMode ? `localhost` : domain
+const expectedOrigin = `https://${rpID}:3001`
```
What's the expected origin exactly? What's running on 3001?

RafaelTaranto commented on new-lamassu-admin/src/pages/Authentication/LoginState.js:
```diff
@@ -107,11 +183,41 @@ const LoginState = ({ state, dispatch }) => {
+                }}
+                buttonClassName={classes.loginButton}
+                className={classes.fidoLoginButtonWrapper}>
+                I have a YubiKey
```
Shouldn't this be a generic text? You can ask Neal for a version of it.

RafaelTaranto commented on new-lamassu-admin/src/pages/Authentication/InputFIDOState.js:
```diff
@@ -0,0 +1,213 @@
+      )}
+      {strategy === 'FIDO2FA' && (
+        <>
+          <H2 className={classes.info}>Insert your Yubikey and touch it</H2>
```
Same issue with "Yubikey" showing here.

chaotixkilla committed:
fix: multiple small fixes

chaotixkilla committed:
fix: expected rp names
